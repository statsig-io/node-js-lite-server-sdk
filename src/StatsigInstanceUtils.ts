import StatsigServer from './StatsigServer';

let _instance: StatsigServer | null = null;

export default abstract class StatsigInstanceUtils {
  static getInstance(): StatsigServer | null {
    return _instance;
  }

  static setInstance(instance: StatsigServer | null) {
    _instance = instance;
  }
}
