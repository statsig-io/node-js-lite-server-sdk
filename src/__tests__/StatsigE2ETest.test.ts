import * as statsigsdk from '../index';
import StatsigInstanceUtils from '../StatsigInstanceUtils';
// @ts-ignore
const statsig = statsigsdk.default;

const CONFIG_SPEC_RESPONSE = JSON.stringify(
  require('./data/download_config_spec.json'),
);

const INIT_RESPONSE = require('./data/initialize_response.json');
let postedLogs = {
  events: [],
};

jest.mock('node-fetch', () => jest.fn());
// @ts-ignore
const fetch = require('node-fetch');
// @ts-ignore
fetch.mockImplementation((url, params) => {
  if (url.includes('download_config_specs')) {
    return Promise.resolve({
      ok: true,
      text: () => Promise.resolve(CONFIG_SPEC_RESPONSE),
    });
  }
  if (url.includes('log_event')) {
    postedLogs = JSON.parse(params.body);
    return Promise.resolve({
      ok: true,
    });
  }
  if (url.includes('get_id_lists')) {
    return Promise.resolve({
      ok: true,
      json: () => Promise.resolve({}),
    });
  }
  if (url.includes('check_gate')) {
    return Promise.resolve({
      ok: true,
      json: () =>
        Promise.resolve({
          name: 'gate_name',
          value: true,
          rule_id: 'fallback_from_server',
        }),
    });
  }
  if (url.includes('get_config')) {
    return Promise.resolve({
      ok: true,
      json: () =>
        Promise.resolve({
          name: 'config_name',
          value: {
            newParam: 'I come from a land down under',
            walk: 10000,
            go: true,
          },
          rule_id: 'fallback_from_server',
        }),
    });
  }
  return Promise.reject();
});

describe('Verify e2e behavior of the SDK with mocked network', () => {
  jest.mock('node-fetch', () => jest.fn());
  const statsigUser = {
    userID: '123',
    email: 'testuser@statsig.com',
  };
  const randomUser = {
    userID: 'random',
    privateAttributes: {
      email: undefined,
    },
  };

  beforeEach(() => {
    jest.restoreAllMocks();
    jest.resetModules();

    StatsigInstanceUtils.setInstance(null);
    postedLogs = {
      events: [],
    };
  });

  test('Verify checkGate and exposure logs', async () => {
    await statsig.initialize('secret-123', { disableDiagnostics: true });
    expect(statsig.getClientInitializeResponse(statsigUser)).toEqual(
      INIT_RESPONSE,
    );
    const on1 = await statsig.checkGate(statsigUser, 'always_on_gate');
    expect(on1).toEqual(true);

    const on2 = await statsig.checkGate(statsigUser, 'always_on_gate');
    expect(on2).toEqual(true);

    const on3 = await statsig.checkGate(statsigUser, 'always_on_gate');
    expect(on3).toEqual(true);

    const passingEmail = await statsig.checkGate(
      statsigUser,
      'on_for_statsig_email',
    );
    expect(passingEmail).toEqual(true);
    const failingEmail = await statsig.checkGate(
      randomUser,
      'on_for_statsig_email',
    );
    expect(failingEmail).toEqual(false);

    // fetch from server does not log an exposure
    const fetchFromServer = await statsig.checkGate(
      statsigUser,
      'fetch_from_server_fallback',
    );
    expect(fetchFromServer).toEqual(true);

    statsig.shutdown();
    expect(postedLogs.events.length).toEqual(3);
    expect(postedLogs.events[0].eventName).toEqual('statsig::gate_exposure');
    expect(postedLogs.events[0].metadata['gate']).toEqual('always_on_gate');
    expect(postedLogs.events[0].metadata['gateValue']).toEqual('true');
    expect(postedLogs.events[0].metadata['ruleID']).toEqual(
      '2DWuOvXQZWKvoaNm27dqcs',
    );

    expect(postedLogs.events[1].eventName).toEqual('statsig::gate_exposure');
    expect(postedLogs.events[1].metadata['gate']).toEqual(
      'on_for_statsig_email',
    );
    expect(postedLogs.events[1].metadata['gateValue']).toEqual('true');
    expect(postedLogs.events[1].metadata['ruleID']).toEqual(
      '3jdTW54SQWbbxFFZJe7wYZ',
    );

    expect(postedLogs.events[2].eventName).toEqual('statsig::gate_exposure');
    expect(postedLogs.events[2].metadata['gate']).toEqual(
      'on_for_statsig_email',
    );
    expect(postedLogs.events[2].metadata['gateValue']).toEqual('false');
    expect(postedLogs.events[2].metadata['ruleID']).toEqual('default');
  });

  test('Verify getConfig and exposure logs', async () => {
    await statsig.initialize('secret-123', { disableDiagnostics: true });
    let config = await statsig.getConfig(statsigUser, 'test_config');
    expect(config.getGroupName()).toBe('statsig emails');
    expect(config.getRuleID()).toBe('4lInPNRUnjUzaWNkEWLFA9');

    expect(config.get('number', 0)).toEqual(7);
    expect(config.get('string', '')).toEqual('statsig');
    expect(config.get('boolean', true)).toEqual(false);
    config = await statsig.getConfig(randomUser, 'test_config');
    expect(config.get('number', 0)).toEqual(4);
    expect(config.get('string', '')).toEqual('default');
    expect(config.get('boolean', false)).toEqual(true);

    // fetch from server does not log an exposure
    config = await statsig.getConfig(statsigUser, 'test_config_fallback');
    expect(config.get('newParam', 'default')).toEqual(
      'I come from a land down under',
    );
    expect(config.get('walk', 10)).toEqual(10000);
    expect(config.get('go', false)).toEqual(true);

    statsig.shutdown();
    expect(postedLogs.events.length).toEqual(2);
    expect(postedLogs.events[0].eventName).toEqual('statsig::config_exposure');
    expect(postedLogs.events[0].metadata['config']).toEqual('test_config');
    expect(postedLogs.events[0].metadata['ruleID']).toEqual(
      '4lInPNRUnjUzaWNkEWLFA9',
    );

    expect(postedLogs.events[1].eventName).toEqual('statsig::config_exposure');
    expect(postedLogs.events[1].metadata['config']).toEqual('test_config');
    expect(postedLogs.events[1].metadata['ruleID']).toEqual('default');
  });

  test('Verify getExperiment and exposure logs', async () => {
    await statsig.initialize('secret-123', { disableDiagnostics: true });
    let experiment = await statsig.getExperiment(
      statsigUser,
      'sample_experiment',
    );
    expect(experiment.get('sample_parameter', true)).toEqual(false);
    experiment = await statsig.getExperiment(randomUser, 'sample_experiment');
    expect(experiment.getGroupName()).toBe('Test');
    expect(experiment.getRuleID()).toBe('5yQbPNUpd8mNbkB0SZZeln');

    expect(experiment.get('sample_parameter', false)).toEqual(true);

    statsig.shutdown();
    expect(postedLogs.events.length).toEqual(2);
    expect(postedLogs.events[0].eventName).toEqual('statsig::config_exposure');
    expect(postedLogs.events[0].metadata['config']).toEqual(
      'sample_experiment',
    );
    expect(postedLogs.events[0].metadata['ruleID']).toEqual(
      '5yQbPMfmKQdiRV35hS3B2l',
    );

    expect(postedLogs.events[1].eventName).toEqual('statsig::config_exposure');
    expect(postedLogs.events[1].metadata['config']).toEqual(
      'sample_experiment',
    );
    expect(postedLogs.events[1].metadata['ruleID']).toEqual(
      '5yQbPNUpd8mNbkB0SZZeln',
    );
  });

  test('Verify getLayer and exposure logs', async () => {
    await statsig.initialize('secret-123', { disableDiagnostics: true });

    // should delegate to a bad config, which fetches from the server
    let layer = await statsig.getLayer(
      statsigUser,
      'd_layer_delegate_to_fallback',
    );
    expect(layer.get('newParam', 'default')).toEqual(
      'I come from a land down under',
    );
    expect(layer.get('walk', 10)).toEqual(10000);
    expect(layer.get('go', false)).toEqual(true);

    statsig.shutdown();
    // fallback does not log an exposure, so nothing gets set here
    expect(postedLogs.events).toEqual([]);
  });

  test('Verify logEvent', async () => {
    await statsig.initialize('secret-123', { disableDiagnostics: true });
    statsig.logEvent(statsigUser, 'add_to_cart', 'SKU_12345', {
      price: '9.99',
      item_name: 'diet_coke_48_pack',
    });
    statsig.shutdown();

    expect(postedLogs.events.length).toEqual(1);
    expect(postedLogs.events[0].eventName).toEqual('add_to_cart');
    expect(postedLogs.events[0].value).toEqual('SKU_12345');
    expect(postedLogs.events[0].metadata['price']).toEqual('9.99');
    expect(postedLogs.events[0].metadata['item_name']).toEqual(
      'diet_coke_48_pack',
    );
    expect(postedLogs.events[0].user.userID).toEqual('123');
    expect(postedLogs.events[0].user.email).toEqual('testuser@statsig.com');
  });
});
