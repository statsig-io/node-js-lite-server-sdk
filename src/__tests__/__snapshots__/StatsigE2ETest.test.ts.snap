// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verify e2e behavior of the SDK with mocked network Verify checkGate and exposure logs 1`] = `
Object {
  "dynamic_configs": Object {
    "86mj1zUC35KaWoOqpe8SKEHQjRlXOzdLfiTJU/iU+vc=": Object {
      "group": "5yQbPMfmKQdiRV35hS3B2l",
      "group_name": "Control",
      "id_type": "userID",
      "is_device_based": false,
      "is_experiment_active": true,
      "is_user_in_experiment": true,
      "name": "86mj1zUC35KaWoOqpe8SKEHQjRlXOzdLfiTJU/iU+vc=",
      "rule_id": "5yQbPMfmKQdiRV35hS3B2l",
      "secondary_exposures": Array [],
      "value": Object {
        "sample_parameter": false,
      },
    },
    "BYdogc4bXxkqfoDOnfjlh1Ij+5QKjNp0rACKdwBL+vc=": Object {
      "group": "",
      "is_device_based": false,
      "is_experiment_active": false,
      "is_user_in_experiment": false,
      "name": "BYdogc4bXxkqfoDOnfjlh1Ij+5QKjNp0rACKdwBL+vc=",
      "rule_id": "",
      "secondary_exposures": Array [],
      "value": Object {},
    },
    "RMv0YJlLOBe7cY7HgZ3Jox34R0Wrk7jLv3DZyBETA7I=": Object {
      "group": "4lInPNRUnjUzaWNkEWLFA9",
      "group_name": "statsig emails",
      "id_type": "userID",
      "is_device_based": false,
      "name": "RMv0YJlLOBe7cY7HgZ3Jox34R0Wrk7jLv3DZyBETA7I=",
      "rule_id": "4lInPNRUnjUzaWNkEWLFA9",
      "secondary_exposures": Array [],
      "value": Object {
        "boolean": false,
        "number": 7,
        "string": "statsig",
      },
    },
    "S52jGlsKiXpNsVbAbe/hcTxLnaHOVT+rNsVP/ibJitY=": Object {
      "group": "prestart",
      "group_name": "Unstarted",
      "id_type": "userID",
      "is_device_based": false,
      "name": "S52jGlsKiXpNsVbAbe/hcTxLnaHOVT+rNsVP/ibJitY=",
      "rule_id": "prestart",
      "secondary_exposures": Array [],
      "value": Object {},
    },
    "n4bQgYhMfWWaL+qgxVrQFaO/TxsrC4Is0V1sFbDwCgg=": Object {
      "group": "prestart",
      "group_name": "Unstarted",
      "id_type": "userID",
      "is_device_based": false,
      "is_experiment_active": false,
      "is_user_in_experiment": false,
      "name": "n4bQgYhMfWWaL+qgxVrQFaO/TxsrC4Is0V1sFbDwCgg=",
      "rule_id": "prestart",
      "secondary_exposures": Array [],
      "value": Object {},
    },
  },
  "evaluated_keys": Object {
    "userID": "123",
  },
  "feature_gates": Object {
    "KfJ1sBBfnL8SwCIxTAyfnj6DiWIyR7uSqimZADQi4Jg=": Object {
      "id_type": "userID",
      "name": "KfJ1sBBfnL8SwCIxTAyfnj6DiWIyR7uSqimZADQi4Jg=",
      "rule_id": "",
      "secondary_exposures": Array [],
      "value": false,
    },
    "OqAaBJ4fpvUSJVicA+jDgotK2OKGJmKxbu+Icyq30P4=": Object {
      "id_type": "userID",
      "name": "OqAaBJ4fpvUSJVicA+jDgotK2OKGJmKxbu+Icyq30P4=",
      "rule_id": "2DWuOvXQZWKvoaNm27dqcs",
      "secondary_exposures": Array [],
      "value": true,
    },
    "rGc+6rvo48V4j1sXkvsGHeSfJfY7kMp1OHfQnw+3XbI=": Object {
      "id_type": "userID",
      "name": "rGc+6rvo48V4j1sXkvsGHeSfJfY7kMp1OHfQnw+3XbI=",
      "rule_id": "2DWuOvXQZWKvoaNm27dqcs",
      "secondary_exposures": Array [],
      "value": true,
    },
    "srhBjpz3NFHATrntxEsdHManNXJQZawtJUkQ3s0XT3Q=": Object {
      "id_type": "userID",
      "name": "srhBjpz3NFHATrntxEsdHManNXJQZawtJUkQ3s0XT3Q=",
      "rule_id": "3jdTW54SQWbbxFFZJe7wYZ",
      "secondary_exposures": Array [],
      "value": true,
    },
  },
  "generator": "statsig-node-lite-sdk",
  "has_updates": true,
  "hash_used": "sha256",
  "layer_configs": Object {
    "FTO1PR918Bsjs1iw6DncMVhpCn/IOS7ZqNYbj2i9Iqg=": Object {
      "explicit_parameters": Array [],
      "group": "default",
      "is_device_based": false,
      "name": "FTO1PR918Bsjs1iw6DncMVhpCn/IOS7ZqNYbj2i9Iqg=",
      "rule_id": "default",
      "secondary_exposures": Array [],
      "undelegated_secondary_exposures": Array [],
      "value": Object {},
    },
    "VHos7km3/fakqhmzEGAt7ATTscBP6wEqSUicn9BX1IM=": Object {
      "allocated_experiment_name": "BYdogc4bXxkqfoDOnfjlh1Ij+5QKjNp0rACKdwBL+vc=",
      "explicit_parameters": Array [],
      "group": "",
      "group_name": "everyone",
      "is_device_based": false,
      "is_experiment_active": false,
      "is_user_in_experiment": false,
      "name": "VHos7km3/fakqhmzEGAt7ATTscBP6wEqSUicn9BX1IM=",
      "rule_id": "",
      "secondary_exposures": Array [],
      "undelegated_secondary_exposures": Array [],
      "value": Object {},
    },
    "yD2Hj5ZefkntKmIwhN9K8Hvdh7helm9hqzp8B9Ng8xc=": Object {
      "allocated_experiment_name": "86mj1zUC35KaWoOqpe8SKEHQjRlXOzdLfiTJU/iU+vc=",
      "explicit_parameters": Array [],
      "group": "5yQbPMfmKQdiRV35hS3B2l",
      "group_name": "Control",
      "is_device_based": false,
      "is_experiment_active": true,
      "is_user_in_experiment": true,
      "name": "yD2Hj5ZefkntKmIwhN9K8Hvdh7helm9hqzp8B9Ng8xc=",
      "rule_id": "5yQbPMfmKQdiRV35hS3B2l",
      "secondary_exposures": Array [],
      "undelegated_secondary_exposures": Array [],
      "value": Object {
        "sample_parameter": false,
      },
    },
  },
  "sdkInfo": ObjectContaining {
    "sdkType": "statsig-node-lite",
    "sdkVersion": Any<String>,
  },
  "sdkParams": Object {},
  "time": Any<Number>,
  "user": Object {
    "email": "testuser@statsig.com",
    "userID": "123",
  },
}
`;
